File: sdks/java/src/main/java/com/blok/nanoservice/nodes/TransformDataNode.java
package com.blok.nanoservice.nodes;

import com.blok.nanoservice.errors.NodeException;
import com.blok.nanoservice.node.NodeHandler;
import com.blok.nanoservice.types.Context;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Transforms JSON data based on field mappings.
 * <p>
 * Config:
 * <ul>
 *   <li>mappings (map, optional) - Field mapping from source to target.
 *       Keys are target field names, values are source field paths (dot-notation).</li>
 *   <li>include_only (list, optional) - If set, only include these fields from output.</li>
 *   <li>exclude (list, optional) - Fields to exclude from output.</li>
 *   <li>defaults (map, optional) - Default values for missing fields.</li>
 * </ul>
 * Input (request body): The input data object to transform.
 * Output: The transformed data object.
 */
public class TransformDataNode implements NodeHandler {

    @Override
    @SuppressWarnings("unchecked")
    public Object execute(Context ctx, Map<String, Object> config) throws Exception {
        Map<String, Object> body = ctx.getRequest().bodyMap();
        if (body == null) {
            throw NodeException.validation("request body must be a JSON object");
        }

        Map<String, Object> result = new HashMap<>();

        // Apply field mappings if configured
        Object mappingsVal = config != null ? config.get("mappings") : null;
        if (mappingsVal instanceof Map<?, ?> mappings) {
            for (Map.Entry<?, ?> entry : mappings.entrySet()) {
                String targetField = String.valueOf(entry.getKey());
                if (entry.getValue() instanceof String sourcePath) {
                    Object value = getNestedValue(body, sourcePath);
                    if (value != null) {
                        result.put(targetField, value);
                    }
                }
            }
        } else {
            // No mappings -- copy all fields
            result.putAll(body);
        }

        // Apply include_only filter
        Object includeOnlyVal = config != null ? config.get("include_only") : null;
        if (includeOnlyVal instanceof List<?> includeOnly && !includeOnly.isEmpty()) {
            Map<String, Object> filtered = new HashMap<>();
            for (Object field : includeOnly) {
                String fieldName = String.valueOf(field);
                if (result.containsKey(fieldName)) {
                    filtered.put(fieldName, result.get(fieldName));
                }
            }
            result = filtered;
        }

        // Apply exclude filter
        Object excludeVal = config != null ? config.get("exclude") : null;
        if (excludeVal instanceof List<?> exclude) {
            for (Object field : exclude) {
                result.remove(String.valueOf(field));
            }
        }

        // Apply defaults for missing fields
        Object defaultsVal = config != null ? config.get("defaults") : null;
        if (defaultsVal instanceof Map<?, ?> defaults) {
            for (Map.Entry<?, ?> entry : defaults.entrySet()) {
                String key = String.valueOf(entry.getKey());
                if (!result.containsKey(key)) {
                    result.put(key, entry.getValue());
                }
            }
        }

        // Store transformed data in vars
        ctx.setVar("transformed_data", result);

        return result;
    }

    /**
     * Retrieves a value from a nested map using dot-notation path.
     *
     * @param data the source map
     * @param path dot-separated field path
     * @return the value at the path, or null if not found
     */
    @SuppressWarnings("unchecked")
    private static Object getNestedValue(Map<String, Object> data, String path) {
        if (data == null || path == null || path.isEmpty()) {
            return null;
        }

        String[] parts = path.split("\\.");
        Object current = data;

        for (String part : parts) {
            if (current instanceof Map<?, ?> map) {
                current = map.get(part);
                if (current == null) {
                    return null;
                }
            } else {
                return null;
            }
        }

        return current;
    }
}

File: sdks/php/examples/TransformDataNode.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Examples;

use Blok\Blok\Errors\NodeException;
use Blok\Blok\Node\NodeHandler;
use Blok\Blok\Types\Context;

/**
 * TransformDataNode transforms JSON data based on field mappings.
 *
 * Config:
 *   - mappings (object, optional): Map of target field name to source field path (dot-notation)
 *   - include_only (array, optional): Only include these fields
 *   - exclude (array, optional): Exclude these fields
 *   - defaults (object, optional): Default values for missing fields
 */
final class TransformDataNode implements NodeHandler
{
    public function execute(Context $ctx, array $config): mixed
    {
        $body = $ctx->request->body;
        if (!is_array($body)) {
            throw NodeException::validation('request body must be a JSON object');
        }

        $result = [];

        // Apply field mappings if configured
        if (isset($config['mappings']) && is_array($config['mappings'])) {
            foreach ($config['mappings'] as $target => $sourcePath) {
                if (is_string($sourcePath)) {
                    $value = $this->getNestedValue($body, $sourcePath);
                    if ($value !== null) {
                        $result[$target] = $value;
                    }
                }
            }
        } else {
            // No mappings - copy all fields
            $result = $body;
        }

        // Apply include_only filter
        if (isset($config['include_only']) && is_array($config['include_only'])) {
            $allowed = array_filter($config['include_only'], 'is_string');
            $result = array_filter(
                $result,
                static fn (string $key): bool => in_array($key, $allowed, true),
                ARRAY_FILTER_USE_KEY,
            );
        }

        // Apply exclude filter
        if (isset($config['exclude']) && is_array($config['exclude'])) {
            foreach ($config['exclude'] as $field) {
                if (is_string($field)) {
                    unset($result[$field]);
                }
            }
        }

        // Apply defaults
        if (isset($config['defaults']) && is_array($config['defaults'])) {
            foreach ($config['defaults'] as $key => $value) {
                if (!array_key_exists($key, $result)) {
                    $result[$key] = $value;
                }
            }
        }

        // Store transformed data in context
        $ctx->setVar('transformed_data', $result);

        return $result;
    }

    /**
     * Get a nested value from an array using dot-notation path.
     */
    private function getNestedValue(array $data, string $path): mixed
    {
        $current = $data;
        foreach (explode('.', $path) as $part) {
            if (!is_array($current) || !array_key_exists($part, $current)) {
                return null;
            }
            $current = $current[$part];
        }
        return $current;
    }
}
